### PR TITLE
[BROWSEUI] -Implement support for shell folder toolbars in the taskbar

### PR DESCRIPTION
**Pull request by yagoulas**

### CODE REVIEW

**Comment by HBelusca** [comment]
It looks like GCC has a nitpickish behaviour:
https://travis-ci.org/reactos/reactos/builds/290007885#L792

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 1]
```diff
@@ -27,20 +27,42 @@ CBandSiteMenu::CBandSiteMenu():
...
+        ILFree(m_QLaunchPidl);
 
     m_BandSite = NULL;
 }
 
+HRESULT WINAPI CBandSiteMenu::FinalConstruct()
+{
+    HRESULT hr = SHGetFolderLocation(0, CSIDL_DESKTOP, NULL, 0, &m_DesktopPidl);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
```
Comment (ThFabba) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: If you fail here, you'll get an uninitialized `m_QLaunchPidl` in the dtor. It's probably best to initialize both pidls to null in the ctor.

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 2]
```diff
@@ -103,12 +125,215 @@ HRESULT CBandSiteMenu::CreateMenuPart()
     return S_OK;
 }
 
+HRESULT CBandSiteMenu::_CreateNewISFBand(HWND hwnd, REFIID riid, void** ppv)
+{
+    WCHAR path[MAX_PATH];
+    WCHAR message[256];
+    BROWSEINFOW bi = { hwnd, NULL, path};
```
Comment (ThFabba) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: Missing space before }

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 3]
```diff
@@ -103,12 +125,215 @@ HRESULT CBandSiteMenu::CreateMenuPart()
...
+        bi.lpszTitle = L"Choose a folder";
+
+    LPITEMIDLIST pidlSelected = SHBrowseForFolderW(&bi);
+    if (pidlSelected == NULL)
+        return S_FALSE;
+
+    CComPtr<IShellFolderBand> pISFB;
+    HRESULT hr = CISFBand_CreateInstance(IID_IShellFolderBand, (PVOID*)&pISFB);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
```
Comment (ThFabba) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: Leaking `pidlSelected` here and below

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 4]
```diff
@@ -103,12 +125,215 @@ HRESULT CBandSiteMenu::CreateMenuPart()
...
+
+    hr = pISFB->InitializeSFB(NULL, pidlSelected);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    return pISFB->QueryInterface(riid, ppv);
```
Comment (ThFabba) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: Actually, even this success path leaks `pidlSelected`

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 5]
```diff
@@ -94,7 +116,7 @@ HRESULT CBandSiteMenu::CreateMenuPart()
         SHGetValue(HKEY_CLASSES_ROOT, wRegKey, NULL, NULL, wszBandName, &dwDataSize);
 
         /* Insert it */
-        InsertMenu(hmenuToolbars, cBands, MF_BYPOSITION, DSA_GetItemCount(m_menuDsa), wszBandName);
+        InsertMenu(hmenuToolbars, cBands, MF_BYPOSITION, DSA_GetItemCount(m_menuDsa) + 0x10, wszBandName);
```
Comment (ThFabba) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: magic 0x10?

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 6]
```diff
@@ -103,12 +125,215 @@ HRESULT CBandSiteMenu::CreateMenuPart()
...
+    BANDINFOSFB bi = {ISFB_MASK_IDLIST};
+    hr = psfb->GetBandInfoSFB(&bi);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return ret;
+
+    LPITEMIDLIST pidl = bi.pidl;
+    if (!pidl)
+        return ret;
+
+    if (pidl->mkid.cb == 0)
```
Comment (ThFabba) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: _ILIsDesktop or such?

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 7]
```diff
@@ -103,12 +125,215 @@ HRESULT CBandSiteMenu::CreateMenuPart()
...
+    ERR("path is %S\n", buffer);
+
+    LPITEMIDLIST pidlQL = ILCreateFromPathW(buffer);
+    if (pidl == NULL)
+        return ret;
+
+    CComPtr<IShellFolder> psfDesktop;
+    hr = SHGetDesktopFolder(&psfDesktop);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return ret;
```
Comment (ThFabba) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: Leaking `pidlQL` here, below, and also in the success case.

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 8]
```diff
@@ -103,12 +125,215 @@ HRESULT CBandSiteMenu::CreateMenuPart()
...
+}
+
+UINT CBandSiteMenu::_GetMenuIdFromBand(CLSID *BandCLSID)
+{
+    /* Try to find the clsid of the band in the dsa */
+    UINT count = DSA_GetItemCount(m_menuDsa);
+    for (UINT i = 0; i < count; i++)
+    {
+        GUID* pdsaGUID = (GUID*)DSA_GetItemPtr(m_menuDsa, i);
+        if (memcmp(pdsaGUID, BandCLSID, sizeof(GUID)) == 0)
```
Comment (ThFabba) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: `IsEqualGUID`?

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 9]
```diff
@@ -103,12 +125,215 @@ HRESULT CBandSiteMenu::CreateMenuPart()
...
+UINT CBandSiteMenu::_GetBandIdFromClsid(CLSID* pclsid)
+{
+    CComPtr<IPersist> pBand;
+    CLSID BandCLSID;
+    DWORD dwBandID;
+    UINT uBand = 0;
+    while (SUCCEEDED(m_BandSite->EnumBands(uBand, &dwBandID)))
+    {
+        if (FAILED(m_BandSite->GetBandObject(dwBandID, IID_PPV_ARG(IPersist, &pBand))))
+            goto nextband;
```
Comment (ThFabba) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: If you make this a for loop you could just use `continue` here

File: dll/win32/browseui/shellbars/CISFBand.cpp [hunk 10]
```diff
@@ -84,7 +88,7 @@ HRESULT CISFBand::CreateSimpleToolbar(HWND hWndParent)
     CComPtr<IEnumIDList> pEndl;    
     LPITEMIDLIST pidl;
     STRRET stret;     
-    hr = m_pISF->EnumObjects(0, SHCONTF_FOLDERS, &pEndl);
+    hr = m_pISF->EnumObjects(0, SHCONTF_FOLDERS|SHCONTF_NONFOLDERS, &pEndl);
```
Comment (ThFabba) [inline dll/win32/browseui/shellbars/CISFBand.cpp]: Isn't an ISFBand supposed to show folders, too?

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 11]
```diff
@@ -120,85 +345,128 @@ HRESULT STDMETHODCALLTYPE CBandSiteMenu::QueryContextMenu(
...
+    idMax = Shell_MergeMenus(hmenu, m_hmenu, indexMenu, idCmdFirst, idCmdLast, MM_DONTREMOVESEPS | MM_SUBMENUSHAVEIDS);
 
     HMENU hmenuToolbars = GetSubMenu(hmenu, indexMenu);
 
     /* Enumerate all present bands and mark them as checked in the menu */
     while (SUCCEEDED(m_BandSite->EnumBands(uBand, &dwBandID)))
     {
         if (FAILED(m_BandSite->GetBandObject(dwBandID, IID_PPV_ARG(IPersist, &pBand))))
-            continue;
+            goto nextband;
```
Comment (ThFabba) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: This fix could make sense as a separate commit if you wanted to split it up.

File: dll/win32/browseui/shellbars/CISFBand.cpp (hunk 10 above)
Comment (yagoulas) [reply dll/win32/browseui/shellbars/CISFBand.cpp]: Yeah, this makes it show files as well.

Comment (ThFabba) [reply dll/win32/browseui/shellbars/CISFBand.cpp]: Oh. Duh. Yeah, I misread that.

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp (hunk 6 above)
Comment (yagoulas) [reply dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: We need to create a helper header and lib at some point for some shared code in various places in the shell including this one.

**Comment by yagoulas** [comment]
I'm going to split this in more meaningful comits later.

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp (hunk 1 above)
Comment (yagoulas) [reply dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: Done

**Comment by yagoulas** [comment]
There is a regression, please don't merge yet.

File: dll/win32/browseui/shellbars/CBandSite.cpp [hunk 12]
```diff
@@ -694,14 +694,37 @@ HRESULT STDMETHODCALLTYPE CBandSiteBase::OnWinEvent(HWND hWnd, UINT uMsg, WPARAM
...
     }
-        
-    if (hWnd == fRebarWindow)
+    else if (uMsg == WM_COMMAND && lParam)
     {
-        /* FIXME: Just send the message? */
-        *plrResult = SendMessageW(hWnd, uMsg, wParam, lParam);
-        return S_OK;
+        hWnd = reinterpret_cast<HWND>(lParam);
+        goto forward_message;
```
Comment (learn-more) [inline dll/win32/browseui/shellbars/CBandSite.cpp]: Not possible to make it a function instead of goto?

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 13]
```diff
@@ -33,14 +44,37 @@ CBandSiteMenu::~CBandSiteMenu()
     if (m_hmenu)
         DestroyMenu(m_hmenu);
 
-    if (m_menuDsa)
-        DSA_Destroy(m_menuDsa);
+    if (m_comcatDsa)
```
Comment (learn-more) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: Maybe a custom `CHeapPtr` with a DSA allocator? (See `CComAllocator`)

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 14]
```diff
@@ -33,14 +44,37 @@ CBandSiteMenu::~CBandSiteMenu()
...
+        DSA_Destroy(m_comcatDsa);
+
+    if (m_DesktopPidl)
+        ILFree(m_DesktopPidl);
```
Comment (learn-more) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: `CComHeapPtr<ITEMIDLIST>`

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 15]
```diff
@@ -94,21 +128,213 @@ HRESULT CBandSiteMenu::CreateMenuPart()
...
     return S_OK;
 }
 
+HRESULT CBandSiteMenu::_CreateNewISFBand(HWND hwnd, REFIID riid, void** ppv)
+{
+    WCHAR path[MAX_PATH];
+    WCHAR message[256];
+    BROWSEINFOW bi = { hwnd, NULL, path };
+
+    if (LoadStringW(GetModuleHandleW(L"browseui.dll"), IDS_BROWSEFORNEWTOOLAR, message, _countof(message)))
```
Comment (learn-more) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: CAtlString has a LoadString function that handles allocation, and saves you from using fixed buffers.

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 16]
```diff
@@ -94,21 +128,213 @@ HRESULT CBandSiteMenu::CreateMenuPart()
...
+        bi.lpszTitle = message;
+    else
+        bi.lpszTitle = L"Choose a folder";
+
+    LPITEMIDLIST pidlSelected = SHBrowseForFolderW(&bi);
```
Comment (learn-more) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: `CComHeapPtr<ITEMIDLIST>`

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 17]
```diff
@@ -94,21 +128,213 @@ HRESULT CBandSiteMenu::CreateMenuPart()
...
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    BANDINFOSFB bi = {ISFB_MASK_IDLIST};
+    hr = psfb->GetBandInfoSFB(&bi);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    if (!bi.pidl)
+        return E_OUTOFMEMORY;
```
Comment (learn-more) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: This silently exits now.

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 18]
```diff
@@ -94,21 +128,213 @@ HRESULT CBandSiteMenu::CreateMenuPart()
...
+    CComPtr<IPersist> pBand;
+    CLSID BandCLSID;
+    DWORD dwBandID;
+
+    for (UINT uBand = 0; SUCCEEDED(m_BandSite->EnumBands(uBand, &dwBandID)); uBand++)
+    {
+        if (FAILED(m_BandSite->GetBandObject(dwBandID, IID_PPV_ARG(IPersist, &pBand))))
+            continue;
+
+        if (FAILED(pBand->GetClassID(&BandCLSID)))
```
Comment (learn-more) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: FAILED_UNEXPECTEDLY?

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp [hunk 19]
```diff
@@ -94,21 +128,213 @@ HRESULT CBandSiteMenu::CreateMenuPart()
...
```
Comment (learn-more) [inline dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: FAILED_UNEXPECTEDLY?

File: dll/win32/browseui/shellbars/CISFBand.h [hunk 20]
```diff
@@ -191,6 +192,4 @@ class CISFBand :
     END_COM_MAP()    
 };
 
-// C Constructor
-extern "C"
-HRESULT WINAPI CISFBand_CreateInstance(REFIID riid, void** ppv);
\ No newline at end of file
+extern "C" HRESULT WINAPI RSHELL_CISFBand_CreateInstance(REFIID riid, void** ppv);
```
Comment (learn-more) [inline dll/win32/browseui/shellbars/CISFBand.h]: Please add a newline at the end of the file :)

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp (hunk 13 above)
Comment (yagoulas) [reply dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: Yeah that would help a lot with freeing pild. I'm not sure that this is applicable to DSAs though (we could just use an ATL list).

File: dll/win32/browseui/shellbars/CBandSite.cpp (hunk 12 above)
Comment (yagoulas) [reply dll/win32/browseui/shellbars/CBandSite.cpp]: Actually I can remove the goto and the behaviour will be the same.

File: dll/win32/browseui/shellbars/CBandSiteMenu.cpp (hunk 15 above)
Comment (yagoulas) [reply dll/win32/browseui/shellbars/CBandSiteMenu.cpp]: I tried to do it but got linker issues. I'm skipping it for now.

**Comment by yagoulas** [comment]
I reorganised the proposed fixes in meaningful commits. Do you guys think that there is something more to think or do you think that I should squash all these changes?
